### PR TITLE
Set the values specified in the Monolog's context in the extra context of the event

### PR DESCRIPTION
**Pull request by ste93cry**
Given the discussion we had in getsentry/sentry-docs#4125 and the fact that this feature is requested repeatedly, let's add it back to the SDK. As a side note, the documentation warns users that the extra context is deprecated, so this is one more reason I strongly believe we should not add this because we cannot know which structured context we would have to create on behalf of the users.

>In addition to structured contexts, Sentry supports adding unstructured "Additional Data" via set_extra. Additional Data is deprecated in favor of structured contexts and should be avoided when possible.

https://docs.sentry.io/platforms/php/enriching-events/context/#additional-data

### CODE REVIEW

**Review by Jean85 — CHANGES_REQUESTED** [review]
Wait, this is not what I was advocating for in the discussion. Making this the default is still not a smart move: my suggestion was to port the handler from the `sentry-laravel` codebase, and leave it unhooked, or put it under an option disabled by default. In that way, we could solve the users' needs, without making it a default behavior of the SDK.

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,7 +17,7 @@
 ## 3.3.0 (2021-05-26)
 
 - Allow setting a custom timestamp on the breadcrumbs (#1193)
-- Add option `ignore_tags` to `IgnoreErrorsIntegration` in order to ignore exceptions by tags values. (#1201)
+- Add option `ignore_tags` to `IgnoreErrorsIntegration` in order to ignore exceptions by tags values (#1201)
```
Comment (glensc) [inline CHANGELOG.md:20]: outside scope of this merge request, move to a separate pull request

**Comment by glensc** [comment]
As noted by @Jean85 and by [me](https://github.com/getsentry/sentry-php/pull/1271#issuecomment-1000749018) this needs to be opt-in, add constructor parameter to the Handler. Otherwise, it's a breaking change.

File: src/Monolog/Handler.php [hunk 2]
```diff
@@ -57,6 +57,15 @@ protected function write(array $record): void
             $scope->setExtra('monolog.channel', $record['channel']);
             $scope->setExtra('monolog.level', $record['level_name']);
 
+            foreach ($record['context'] as $key => $value) {
```
Comment (Wick96) [inline src/Monolog/Handler.php:60]: @glensc Hi i have two ideas :)

The first one is to maybe just add logger context as  `$scope->setExtra('monolog.context', $record['context']);` and then the user can freely filter or anonymize this array later in `before_send` hook if he needs it.

Sec idea... maybe for now just add context to hint `$hint->extra = ['monolog.context', $record['context']];` so the user can then manually filter & add data he needs to event in `before_send` hook.

Just "my two cents".

File: src/EventId.php [hunk 3]
```diff
@@ -35,7 +35,7 @@ public function __construct(string $value)
      */
     public static function generate(): self
     {
-        return new self(str_replace('-', '', uuid_create(UUID_TYPE_RANDOM)));
+        return new self(str_replace('-', '', uuid_create(\UUID_TYPE_RANDOM)));
```
Comment (glensc) [inline src/EventId.php:38]: unrelated change to the pull request. should be submitted separately

File: CHANGELOG.md (hunk 1 above)
Comment (ste93cry) [reply CHANGELOG.md:31]: I'm just removing the final dot to keep consistency with the rest of the points of the list, it's not worth opening a separate PR for this 😅

Comment (glensc) [reply CHANGELOG.md:31]: it's not about worthiness, it's about quality. you don't stuff unrelated changes together.

please read about atomic commits:
- https://www.freshconsulting.com/insights/blog/atomic-commits/

so, now consider your changeset needs to be reverted, but since it's all same commit, your unrelated change will be reverted too with `git revert`.

also, unrelated noise makes PR more readable and the style change PR gets merged very likely sooner as it's not affecting functionality.

File: src/EventId.php (hunk 3 above)
Comment (glensc) [reply src/EventId.php:38]: and since already nit-picking about commits quality, having commit titled "CR fixes" does not give any value to commit history, since you already rebased and force pushed, squash it with original commit.

File: CHANGELOG.md (hunk 1 above)
Comment (ste93cry) [reply CHANGELOG.md:31]: I understand your motivation and I agree with your reasoning when talking about big changes not related to the PR's purpose, but commenting that adding a dot to a phrase is something that should be done separately has no value for the code review in my opinion and it's just nitpicking

File: src/EventId.php (hunk 3 above)
Comment (ste93cry) [reply src/EventId.php]: >should be submitted separately

This change actually is unwanted, it seems that when `ext-uuid` extension is installed (like on my environment) `PHP-CS-Fixer` consider the constant as available natively while when not present (but the polyfill is) it doesn't.

File: src/Monolog/Handler.php (hunk 2 above)
Comment (ste93cry) [reply src/Monolog/Handler.php]: @Jean85 @stayallive opinions about the first point? I'm kind of undecided if we should put everything under the extra context or if as, suggested, it would be wiser to group everything under monolog.context

Comment (Jean85) [reply src/Monolog/Handler.php]: Maybe the second option, a dedicated context, would be better? That way we could avoid situation where we accidentally override user-submitted data.

Comment (ste93cry) [reply src/Monolog/Handler.php]: Done!

Comment (glensc) [reply src/Monolog/Handler.php]: @ste93cry if you inow how to rebase and force push, why don't you already cleanup the history? like why have commit stuff, revert stuff? just squash so that it's a commit that adds stuff with the reverted stuff in same commit?

Comment (ste93cry) [reply src/Monolog/Handler.php]: Keeping the history of commits in the PR is sometimes useul to keep track of the process that led to the final version of the code to review, giving more context to the reviewer. Besides that, squashing is done when merging into the target branch so it doesn't really matter how many commits there are here, the history of the repository will be clean anyway :wink:

_Funny fact: you were the one advocating for commits to be as atomic as possible so that it's possible to revert them without losing other unrelated changes_

Comment (glensc) [reply src/Monolog/Handler.php]: That completely changes the story if the pull request is squash merged. I wish github/gitlab present it to reviewers that squash merge is used, then could just omit any commit-related quality checks.

as for the atomic part, for the purposes of merging cleaner history, I would exclude (with rebase) changes from merge that are removed rather than merging revert commits. as everything before merge can be rewritten. if you want to preserve historic decisions, then maybe start with new pull request leave old one for posterity.

so far both github/gitlab still don't present well previous versions of pushes. gitlab at least has merge request versions, but in github you have to dig through timeline find the "has force pushed" and then can find previous versions.

**Review by Jean85 — APPROVED** [review]

**Review by stayallive — APPROVED** [review]
Looks good 👍
